BrowserViewportSevice: Rework of IBreakpointService & IResizeService

ScarletKuro opened the pull request:
## Description
Replaces `IBreakpointService` and `IResizeService`
Reasons: https://github.com/MudBlazor/MudBlazor/pull/7074#issuecomment-1605125832

**Main API:**

<details>
  <summary>ResizeOptions API (old, reused)</summary>
  
```CSharp
/// <summary>
/// Represents options for <see cref="IBreakpointService"/>.
/// </summary>
public class ResizeOptions : IEquatable<ResizeOptions>
{
	/// <summary>
	/// Rate in milliseconds that the browsers `resize()` event should report a change.
	/// Setting this value too low can cause poor application performance.
	/// Default value is <c>100</c>.
	/// </summary>
	public int ReportRate { get; set; } = 100;

	/// <summary>
	/// Report resize events and media queries in the browser's console.
	/// Default value is <c>false</c>.
	/// </summary>
	public bool EnableLogging { get; set; } = false;

	/// <summary>
	/// Suppress the first OnResized that is invoked when a new event handler is added.
	/// Default value is <c>true</c>.
	/// </summary>
	public bool SuppressInitEvent { get; set; } = true;

	/// <summary>
	/// If true, RaiseOnResized is called only when breakpoint has changed.
	/// Default value is <c>true</c>.
	/// </summary>
	public bool NotifyOnBreakpointOnly { get; set; } = true;

	/// <summary>
	/// Gets or sets the breakpoint definitions, representing specific breakpoints and their associated width.
	/// </summary>
	/// <value>
	/// A dictionary where each entry represents a breakpoint, and the corresponding <c>int</c> value represents the width.
	/// </value>
	/// <remarks>
	/// When the dictionary is null or empty, the default breakpoint definitions will be used.
	/// The default breakpoint definitions are as follows:
	/// [Breakpoint.Xxl] = 2560,
	/// [Breakpoint.Xl] = 1920,
	/// [Breakpoint.Lg] = 1280,
	/// [Breakpoint.Md] = 960,
	/// [Breakpoint.Sm] = 600,
	/// [Breakpoint.Xs] = 0.
	/// </remarks>
	public Dictionary<Breakpoint, int>? BreakpointDefinitions { get; set; } = new();
}
  ```
  
</details>

<details>
  <summary>IBrowserViewportService API (new)</summary>
  
```CSharp
/// <summary>
/// Represents a service that serves to listen to browser window size changes and breakpoints.
/// </summary>
/// <remarks>
/// This service replaces <see cref="IBreakpointService"/>, <see cref="IResizeService"/> and <see cref="IResizeListenerService"/>.
/// </remarks>
public interface IBrowserViewportService : IAsyncDisposable
{
    /// <summary>
    /// Gets the current resize options.
    /// </summary>
    ResizeOptions ResizeOptions { get; }

    /// <summary>
    /// Subscribes an observer to receive notifications of browser window size changes and breakpoints.
    /// </summary>
    /// <param name="observer">The observer to subscribe.</param>
    /// <param name="fireImmediately">Indicates whether the event will fire immediately with the current <see cref="BrowserWindowSize"/> and <see cref="Breakpoint"/> information without waiting for changes.
    /// When set to <c>true</c>, the event will be fired immediately. When set to <c>false</c>, it will wait for the service to observe any changes before firing the event.
    /// The event <see cref="IBrowserViewportObserver.NotifyBrowserViewportChangeAsync"/> won't fire if such <see cref="IBrowserViewportObserver.Id"/> already exist, for example when you re-subscribe.
    /// </param>
    /// <returns>A task representing the asynchronous operation.</returns>
    Task SubscribeAsync(IBrowserViewportObserver observer, bool fireImmediately = true);

    /// <summary>
    /// Subscribes a lambda <see cref="Action{BrowserViewportEventArgs}"/> with a unique ID to receive notifications of browser window size changes and breakpoints.
    /// </summary>
    /// <param name="observerId">The unique ID associated with the observer. Use this ID to later <see cref="UnsubscribeAsync(Guid)"/>.</param>
    /// <param name="lambda">The lambda function to subscribe.</param>
    /// <param name="options">The resize options for the observer. When set to null, the global options provided during AddMudServices/AddMudBlazorResizeListener will be used.
    /// When specific options are provided, they will be used to observe the changes.
    /// The <see cref="Action{BrowserViewportEventArgs}"/> won't be invoked if such <see cref="IBrowserViewportObserver.Id"/> already exist, for example when you re-subscribe.
    /// After you pass the options, modifying the instance won't have any effect, including re-subscription, as C# and JS side doesn't support this, you need to <see cref="UnsubscribeAsync(Guid)"/> and subscribe again.
    /// </param>
    /// <param name="fireImmediately">Indicates whether the event will fire immediately with the current <see cref="BrowserWindowSize"/> and <see cref="Breakpoint"/> information without waiting for changes. When set to <c>true</c>, the event will be fired immediately. When set to <c>false</c>, it will wait for the service to observe any changes before firing the event.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    Task SubscribeAsync(Guid observerId, Action<BrowserViewportEventArgs> lambda, ResizeOptions? options = null, bool fireImmediately = true);

    /// <summary>
    /// Subscribes a lambda <see cref="Func{BrowserViewportEventArgs, Task}"/> with a unique ID to receive notifications of browser window size changes and breakpoints.
    /// </summary>
    /// <param name="observerId">The unique ID associated with the observer. Use this ID to later <see cref="UnsubscribeAsync(Guid)"/>.</param>
    /// <param name="lambda">The lambda function to subscribe.</param>
    /// <param name="options">The resize options for the observer. When set to null, the global options provided during AddMudServices/AddMudBlazorResizeListener will be used.
    /// When specific options are provided, they will be used to observe the changes.
    /// The <see cref="Func{BrowserViewportEventArgs, Task}"/> won't be invoked if such <see cref="IBrowserViewportObserver.Id"/> already exist, for example when you re-subscribe.
    /// After you pass the options, modifying the instance won't have any effect, including re-subscription, as C# and JS side doesn't support this, you need to <see cref="UnsubscribeAsync(Guid)"/> and subscribe again.
    /// </param>
    /// <param name="fireImmediately">Indicates whether the event will fire immediately with the current <see cref="BrowserWindowSize"/> and <see cref="Breakpoint"/> information without waiting for changes. When set to <c>true</c>, the event will be fired immediately. When set to <c>false</c>, it will wait for the service to observe any changes before firing the event.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    Task SubscribeAsync(Guid observerId, Func<BrowserViewportEventArgs, Task> lambda, ResizeOptions? options = null, bool fireImmediately = true);

    /// <summary>
    /// Unsubscribes an observer from receiving notifications.
    /// </summary>
    /// <param name="observer">The observer to unsubscribe.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    Task UnsubscribeAsync(IBrowserViewportObserver observer);

    /// <summary>
    /// Unsubscribes with the specified ID from receiving notifications.
    /// </summary>
    /// <param name="observerId">The unique ID associated with the observer to unsubscribe.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    Task UnsubscribeAsync(Guid observerId);

    /// <summary>
    /// Matches if the document currently matches the media query, or false if not. 
    /// </summary>
    /// <param name="mediaQuery">A string specifying the media query.</param>
    /// <returns>A task representing a boolean value that is <c>true</c> if the document currently matches the media query; otherwise, it's <c>false</c>.</returns>
    Task<bool> IsMediaQueryMatchAsync(string mediaQuery);

    /// <summary>
    /// Check if the current breakpoint fits within the current window size
    /// </summary>
    /// <param name="breakpoint">The breakpoint to check.</param>
    /// <returns>A task representing whether the media size meets the criteria. Returns <c>true</c> if the media size meets the criteria; otherwise, returns <c>false</c>. For example, if the current window size is <see cref="Breakpoint.Sm"/> and the breakpoint is set to <see cref="Breakpoint.SmAndDown"/>, this method will return <c>true</c>.</returns>
    Task<bool> IsBreakpointWithinWindowSizeAsync(Breakpoint breakpoint);

    /// <summary>
    /// Check if the current breakpoint fits within the reference size
    /// </summary>
    /// <param name="breakpoint">The breakpoint to check</param>
    /// <param name="reference">The reference breakpoint (<see cref="Breakpoint.Xs"/>, <see cref="Breakpoint.Sm"/>, <see cref="Breakpoint.Md"/>, <see cref="Breakpoint.Lg"/>,<see cref="Breakpoint.Xl"/>, <see cref="Breakpoint.Xxl"/>)</param>
    /// <returns>A task representing whether the media size meets the criteria. Returns <c>true</c> if the media size meets the criteria; otherwise, returns <c>false</c>. For example, if the reference size is <see cref="Breakpoint.Sm"/> and the breakpoint is set to <see cref="Breakpoint.SmAndDown"/>, this method will return <c>true</c>.</returns>
    Task<bool> IsBreakpointWithinReferenceSizeAsync(Breakpoint breakpoint, Breakpoint reference);

    /// <summary>
    /// Get the current breakpoint
    /// </summary>
    /// <returns>A task representing the current breakpoint</returns>
    Task<Breakpoint> GetCurrentBreakpointAsync();

    /// <summary>
    /// Get the current size of the window
    /// </summary>
    /// <returns>A task representing the current browser size</returns>
    Task<BrowserWindowSize> GetCurrentBrowserWindowSizeAsync();
}
```
  
</details>

<details>
  <summary>IBrowserViewportObserver API (new)</summary>
  
```CSharp
/// <summary>
/// Represents an observer for browser viewport updates.
/// </summary>
public interface IBrowserViewportObserver
{
    /// <summary>
    /// Gets the unique identifier of the observer.
    /// </summary>
    Guid Id { get; }

    /// <summary>
    /// Gets the resize options for the observer.
    /// When set to null, the global options provided during AddMudServices / AddMudBlazorResizeListener will be used.
    /// When specific options are provided, they will be used to observe the changes.
    /// </summary>
    /// <remarks>
    /// After you set the options, modifying the instance won't have any effect, including re-subscription, as C# and JS side doesn't support this, you need to <see cref="IBrowserViewportService.UnsubscribeAsync(IBrowserViewportObserver)"/> and subscribe again.
    /// </remarks>
    ResizeOptions? ResizeOptions => null;

    /// <summary>
    /// Notifies the observer of browser size and breakpoint change.
    /// </summary>
    /// <param name="browserViewportEventArgs">The event arguments containing information about the <see cref="BrowserWindowSize"/> and <see cref="Breakpoint"/> change.</param>
    /// <returns>A task representing the asynchronous notification operation.</returns>
    Task NotifyBrowserViewportChangeAsync(BrowserViewportEventArgs browserViewportEventArgs);
}
  ```
  
</details>

<details>
  <summary>BrowserWindowSize API (old, reused)</summary>
  
```CSharp
/// <summary>
/// Represents the size of a browser window.
/// </summary>
public class BrowserWindowSize : EventArgs
{
    /// <summary>
    /// Gets or sets the height of the browser window.
    /// </summary>
    public int Height { get; set; }

    /// <summary>
    /// Gets or sets the width of the browser window.
    /// </summary>
    public int Width { get; set; }
}
  ```
  
</details>

<details>
  <summary>BrowserViewportEventArgs API (new)</summary>
  
```CSharp
/// <summary>
/// Provides data for the browser viewport event.
/// </summary>
public class BrowserViewportEventArgs : EventArgs
{
    /// <summary>
    /// Gets the ID of the JavaScript listener.
    /// </summary>
    public Guid JavaScriptListenerId { get; }

    /// <summary>
    /// Gets the browser window size.
    /// </summary>
    public BrowserWindowSize BrowserWindowSize { get; }

    /// <summary>
    /// Gets the breakpoint associated with the browser size.
    /// </summary>
    public Breakpoint Breakpoint { get; }

    /// <summary>
    /// Gets a value indicating whether this is the first event that was fired.
    /// This is true when you set <c>fireImmediately</c> to <c>true</c> in the <see cref="IBrowserViewportService.SubscribeAsync(IBrowserViewportObserver, bool)"/>, <see cref="IBrowserViewportService.SubscribeAsync(Guid, Action{BrowserViewportEventArgs}, ResizeOptions?, bool)"/>, <see cref="IBrowserViewportService.SubscribeAsync(Guid, Func{BrowserViewportEventArgs, Task}, ResizeOptions?, bool)"/>  method.
    /// </summary>
    public bool IsImmediate { get; }

    /// <summary>
    /// Initializes a new instance of the <see cref="BrowserViewportEventArgs"/> class.
    /// </summary>
    /// <param name="javaScriptListenerId">The ID of the JavaScript listener.</param>
    /// <param name="browserWindowSize">The browser window size.</param>
    /// <param name="breakpoint">The breakpoint associated with the browser size.</param>
    /// <param name="isImmediate">Specifies whether this is the first event that was fired.</param>
    public BrowserViewportEventArgs(Guid javaScriptListenerId, BrowserWindowSize browserWindowSize, Breakpoint breakpoint, bool isImmediate = false);
}
  ```
  
</details>

**Usage example:**

</details>

<details>
  <summary>Using lambda (Func or Action)</summary>
  
```CSharp
[Inject]
private IBrowserViewportService BrowserViewportService { get; set; }

private int _width = 0;
private int _height = 0;

private Guid _subscriptionId = Guid.NewGuid();

protected override async Task OnAfterRenderAsync(bool firstRender)
{
	if (firstRender)
	{
		await BrowserViewportService.SubscribeAsync(_subscriptionId, Lambda, fireImmediately: true);
	}

	await base.OnAfterRenderAsync(firstRender);
}

public async ValueTask DisposeAsync() => await BrowserViewportService.UnsubscribeAsync(_subscriptionId);

private async Task Lambda(BrowserViewportEventArgs args)
{
	_width = args.BrowserWindowSize.Width;
	_height = args.BrowserWindowSize.Height;

	await InvokeAsync(StateHasChanged);
}
  ```
  
</details>

</details>

<details>
  <summary>Using IBrowserViewportObserver</summary>
  
```CSharp
//@implements IBrowserViewportObserver

[Inject]
private IBrowserViewportService BrowserViewportService { get; set; }

private int _width = 0;
private int _height = 0;

protected override async Task OnAfterRenderAsync(bool firstRender)
{
	if (firstRender)
	{
		await BrowserViewportService.SubscribeAsync(this, fireImmediately: true);
	}

	await base.OnAfterRenderAsync(firstRender);
}

public async ValueTask DisposeAsync() => await BrowserViewportService.UnsubscribeAsync(this);

Guid IBrowserViewportObserver.Id { get; } = Guid.NewGuid();

ResizeOptions IBrowserViewportObserver.ResizeOptions { get; } = new()
{
	ReportRate = 50,
	NotifyOnBreakpointOnly = false
};

Task IBrowserViewportObserver.BrowserViewportChangeNotificationAsync(BrowserViewportEventArgs browserViewportEventArgs)
{
	_width = browserViewportEventArgs.BrowserWindowSize.Width;
	_height = browserViewportEventArgs.BrowserWindowSize.Height;

	return InvokeAsync(StateHasChanged);
}
  ```
  
</details>

## How Has This Been Tested?
Updates existing bUnit tests and added many new NUnit tests.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

ScarletKuro committed:
BrowserViewportSevice: Rework of IBreakpointService & IResizeService

ScarletKuro commented:
@henon still work in progress, missing a lot of polishing, test coverage and xml comments, but it's already a working prototype integrated in `MudHidden` and `MudBreakpointProvider`

ScarletKuro commented:
I have implemented a similar optimization as the previous implementation. The original approach avoids creating a new JavaScript "listener" every time a component subscribes. Instead, it checks if a listener with the corresponding `ResizeOption` already exists, and only creates a new one if necessary.

You can have a scenario where you might have 10 observers monitoring changes, but only one JavaScript listener on the other side. In the previous implementation, the number of observers and JavaScript listeners would have been equal.

Implementing this optimization does add complexity to the code, but with proper documentation, I believe I can explain how it works and why it's implemented this way.

ScarletKuro commented:
Fixing the `HiddenTests` will take me a lot of time. The original author mocked the `IBreakpointService`, which I believe is not a good idea. You are suppose to mock components that cannot be executed in tests, i.e. `IJsRuntime`.

Another problem with mocking the `IBreakpointService` is that if there are any changes made to the actual `BreakpointService` implementation, the tests may give a false sense of security, as the mock would still function as before without reflecting the real behavior.

For example, in one of the tests, there is a line:
`listenerMock.Setup(x => x.IsMediaSize(Breakpoint.Lg, Breakpoint.Xs)).Returns(true).Verifiable();`
However, in the real implementation, this condition would never be true, and there is no configuration for that specific method:
https://github.com/MudBlazor/MudBlazor/blob/30cd8b836bc0444327b3032bb7100bfa4799a55c/src/MudBlazor/Services/ResizeListener/BreakpointService%20.cs#L100
For me personally it means the test is not really correct.
Fixing one test took me almost whole night, since I need to confirm that the problem lies within the test and not my own. Then, I must validate that the test is indeed incorrect and ensure that my test is accurate, while ensuring the rest of the functionality is functioning as expected etc :disappointed_relieved:

But since I don't mock `IBrowserViewportService`, I can't test how much times was the XYZ method called, like it does in the test, but I don't know how much is this relevant considering that the end result is correct i.e. that mudhidden is visible or not depending on the conditions.

henon commented:
Benno took testing to an extreme, so I don't mind if you relax the one or other test. As you said, how many times a function is called is not really that relevant and I am pretty sure that you won't code a service that calls one method hundreds of times ;) 

You can change or deactivate those test which are too much work to get working with the new implementation.

ScarletKuro commented:
> Benno took testing to an extreme, so I don't mind if you relax the one or other test. As you said, how many times a function is called is not really that relevant and I am pretty sure that you won't code a service that calls one method hundreds of times ;)
> 
> You can change or deactivate those test which are too much work to get working with the new implementation.

Yeah, I manually counter on few tests to see that it's pretty similar.
I have made the tests working now, just not doing any method counts.

Now just need to cover with NUnit BrowserViewportSevice without any BUnit and component like I did with popover, change docs, add few comments and it will be ready.

ScarletKuro commented:
Coverage is 100% now, tho I still want to add some explicit tests like for `UnsubscribeAsync`, and the `CreateJavaScriptListener` how it create listener only per different options and then I can say that major work is done.

ScarletKuro commented:
@henon it's done now

ScarletKuro commented:
Tho, i would like to somehow to add this code just as snippet to `BrowserResizeEventExample.razor` without running the example code:

```
    [Inject]
    private IBrowserViewportService BrowserViewportService { get; set; }

    private int _width = 0;
    private int _height = 0;

    private Guid _subscriptionId = Guid.NewGuid();

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            await BrowserViewportService.SubscribeAsync(_subscriptionId, async args =>
            {
                _width = args.BrowserWindowSize.Width;
                _height = args.BrowserWindowSize.Height;

                await InvokeAsync(StateHasChanged);
            });
        }

        await base.OnAfterRenderAsync(firstRender);
    }

    public async ValueTask DisposeAsync() => await BrowserViewportService.UnsubscribeAsync(_subscriptionId);
```

I just want to add that if they don't want to implement `IBrowserViewportObserver` like it shown in current example, then then can alternatively use an Action or Func lambda and thats how it looks.

ScarletKuro commented:
I will later add the API like in the popover service to the header later

ScarletKuro commented:
Didn't replace with new service in the `MudDrawer` and `BreakpointListenerExample`

ScarletKuro commented:
I addressed everything and migrated the left overs.
The coverage slightly dropped because on of `MudDrawer`, coverate think it's new lines but in reality I just replaced to new method names.
Tbh I did not expect this to be so massive... have fun reviewing.
I tried to test visually as much as possible. Unit test wise I tried to cover everything I could, it's also well documented compared to old implementations, earlier I couldn't even figure out what they doing.
The only testing I haven't done yet is the testing on BSS and WinForm/WP, but in general it's ready for review.
P.S I updated the head with API description and usage examples

ScarletKuro commented:
I also want to add a note about the `SuppressInitEvent` in the `ResizeOption` that existed before me and the `bool fireImmediately = true` parameter in the `SubscribeAsync` method of my service. 

The difference between them:
The main difference between `SuppressInitEvent` and `fireImmediately` is their implementation. `SuppressInitEvent` is handled on the JavaScript side, while `fireImmediately` is purely implemented in C#. Despite their different implementations, they serve the same purpose: firing the event immediately with the initial breakpoint and browser size once you have subscribed, rather than waiting for a potential browser size change that may never occur.

Another difference is the default settings. `SuppressInitEvent` is disabled by default, while `fireImmediately` is enabled by default. The reason behind this is that the previous service had a return value in the subscribe method with this data, which is considered a bad practice when using the observer pattern. To align with best practices, I removed the return value and made `fireImmediately` default to `true`. Additionally, I added a property called `IsImmediate` in the callback, which allows you to check if the current event is the initial event. This property is essential when you need to perform specific actions based on the initial event, as demonstrated in the following code snippet:

```CSharp
Task IBrowserViewportObserver.BrowserViewportChangeNotificationAsync(BrowserViewportEventArgs browserViewportEventArgs)
{
    if (browserViewportEventArgs.IsImmediate)
    {
        _start = browserViewportEventArgs.Breakpoint;
    }
    else
    {
        _breakpointHistory.Add(browserViewportEventArgs.Breakpoint);
    }

    return InvokeAsync(StateHasChanged);
}
```

You may wonder why I didn't reuse the existing `SuppressInitEvent`. Firstly, reusing it would not allow us to differentiate the initial event using the `IsImmediate` property. Although we could modify the original JavaScript code to include this information, the main reason for not reusing it is that we cannot change the default value from `true` to `false` due to backward compatibility concerns. Changing the default behavior could lead to unexpected changes for users who still rely on the old `BreakpointService` / `ResizeService` / `ResizeListenerService`.

In the future, when we remove the old service, I plan to remove the `SuppressInitEvent` property from `ResizeOptions` altogether.

henon commented:
I have the following suggestions for the API naming:

1) Rename `Task<bool> IsMatchMediaAsync(string mediaQuery);` to `IsMediaMatchAsync` or even `IsMediaQueryMatchAsync` or `MatchesMediaQueryAsync`. 

2) Rename `Task BrowserViewportChangeNotificationAsync(BrowserViewportEventArgs browserViewportEventArgs);` to either `NotifyBrowserViewportChange` or `FireBrowserViewportChangeNotification`

The rest looks good.

ScarletKuro commented:
>     1. Rename `Task<bool> IsMatchMediaAsync(string mediaQuery);` to `IsMediaMatchAsync` or even `IsMediaQueryMatchAsync` or `MatchesMediaQueryAsync`.


I like the `IsMediaQueryMatchAsync` because it aligns with https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia

>     2. Rename `Task BrowserViewportChangeNotificationAsync(BrowserViewportEventArgs browserViewportEventArgs);` to either `NotifyBrowserViewportChange` or `FireBrowserViewportChangeNotification`

NotifyBrowserViewportChangeAsync works, it's just that in popover we named `PopoverCollectionUpdatedNotificationAsync`

henon commented:
> `NotifyBrowserViewportChangeAsync` works, it's just that in popover we named `PopoverCollectionUpdatedNotificationAsync`

Yeah, `PopoverCollectionUpdatedNotificationAsync` is an unfortunat name for a method. It's like calling a method that adds something `Addition` ;)

ScarletKuro committed:
Rename

ScarletKuro commented:
Done with renaming

ScarletKuro commented:
Hi @tjscience
This is revamped service that combines the functionalities of `IBreakpointService`, `IResizeService`, and `IResizeListenerService`.
It support for the async `Func<BrowserWindowSize, Task>` that you have requested:

> Anyone familiar with ResizeService? Why was SubscribeAsync and UnsubscribeAsync written with Action<BrowserWindowSize> instead of Func<BrowserWindowSize, Task>? This prevents the ability to (easily) have async in the callback.

See example:
</details>

<details>
  <summary>Using lambda (Func or Action)</summary>
  
```CSharp
[Inject]
private IBrowserViewportService BrowserViewportService { get; set; }

private int _width = 0;
private int _height = 0;

private Guid _subscriptionId = Guid.NewGuid();

protected override async Task OnAfterRenderAsync(bool firstRender)
{
	if (firstRender)
	{
		await BrowserViewportService.SubscribeAsync(_subscriptionId, Lambda, fireImmediately: true);
	}

	await base.OnAfterRenderAsync(firstRender);
}

public async ValueTask DisposeAsync() => await BrowserViewportService.UnsubscribeAsync(_subscriptionId);

private async Task Lambda(BrowserViewportEventArgs args)
{
	_width = args.BrowserWindowSize.Width;
	_height = args.BrowserWindowSize.Height;

	await InvokeAsync(StateHasChanged);
}
  ```
  
</details>

</details>

<details>
  <summary>Using IBrowserViewportObserver</summary>
  
```CSharp
//@implements IBrowserViewportObserver

[Inject]
private IBrowserViewportService BrowserViewportService { get; set; }

private int _width = 0;
private int _height = 0;

protected override async Task OnAfterRenderAsync(bool firstRender)
{
	if (firstRender)
	{
		await BrowserViewportService.SubscribeAsync(this, fireImmediately: true);
	}

	await base.OnAfterRenderAsync(firstRender);
}

public async ValueTask DisposeAsync() => await BrowserViewportService.UnsubscribeAsync(this);

Guid IBrowserViewportObserver.Id { get; } = Guid.NewGuid();

ResizeOptions IBrowserViewportObserver.ResizeOptions { get; } = new()
{
	ReportRate = 50,
	NotifyOnBreakpointOnly = false
};

Task IBrowserViewportObserver.BrowserViewportChangeNotificationAsync(BrowserViewportEventArgs browserViewportEventArgs)
{
	_width = browserViewportEventArgs.BrowserWindowSize.Width;
	_height = browserViewportEventArgs.BrowserWindowSize.Height;

	return InvokeAsync(StateHasChanged);
}
  ```
  
</details>

We wanted to ask if you need this already in v6, or we can more this to the new v7 that has currently no ETA.

ScarletKuro commented:
On the other hand, having this in v6 as well gives benefit. In next PR we can mark the old ones as obsolete and let people know, and it would be less overwhelming migration changes from v6 to v7 in future?

henon commented:
That is a really good argument. OK. merging now

henon commented:
Many thanks for all the hard work @ScarletKuro, I know this cost you a couple of nights.

tjscience commented:
This is awesome

> 

Really great work on this @ScarletKuro! I could definitely utilize this in v6. Awesome!

ScarletKuro commented:
Forgot to tell, that I have also tested on BSS and WinForm, and didn't find any problems.

Yomodo commented:
Thanks for your amazing work @ScarletKuro !
